privval/src/server.rs

```diff
@@ -0,0 +1,467 @@
+        chain_id,
+        new_state.block_id_prefix(),
+        new_state,
+        started_at.elapsed().as_millis(),
+    );
+    if let Err(e) = state_provider.persist_state(&new_state).await {
+        error!("[{}] failed to persist the state: {}", chain_id, e);
+    }
+    Ok(signature)
+}
+
+/// raw error types for the responses
+fn get_invalid_chain_id_error(chain_id: &chain::Id) -> RemoteSignerError {
+    RemoteSignerError {
+        code: 1,
+        description: format!("invalid chain id: {}", chain_id),
+    }
+}
+
+/// raw error types for the responses
+fn get_double_sign_error(height: &block::Height) -> RemoteSignerError {
+    RemoteSignerError {
+        code: 2,
+        description: format!("double signing requested at height: {}", height),
+    }
+}
+
+/// raw error types for the responses
+fn get_state_not_found_error() -> RemoteSignerError {
+    RemoteSignerError {
+        code: 3,
+        description: "existing state failed to load (internal error)".to_owned(),
+    }
+}
+
+/// raw error types for the responses
+fn get_failed_to_sign_error() -> RemoteSignerError {
+    RemoteSignerError {
+        code: 4,
+        description: "signer failed to sign (internal error)".to_owned(),
+    }
+}
+
+#[tonic::async_trait]
+impl<S, VS> PrivValidatorApi for PrivvalService<S, VS>
+where
+    S: SignerProvider + 'static,
+    VS: ValidatorStateProvider + Sync + Send + 'static,
+{
+    async fn get_pub_key(
+        &self,
+        request: Request<RawPubKeyRequest>,
+    ) -> Result<Response<RawPubKeyResponse>, Status> {
+        let req = PubKeyRequest::try_from(request.into_inner())
+            .map_err(|e| Status::invalid_argument(e.to_string()))?;
+        debug!("[{}] received a public key request", req.chain_id);
+        let resp = match self.pubkeys.get(&req.chain_id) {
+            Some(pubkey) => PubKeyResponse {
+                pub_key: Some(*pubkey),
+                error: None,
+            },
+            None => {
+                error!("[{}] no public key found", req.chain_id);
+                PubKeyResponse {
+                    pub_key: None,
+                    error: Some(get_invalid_chain_id_error(&req.chain_id)),
+                }
+            },
+        };
+        Ok(Response::new(resp.into()))
+    }
+    async fn sign_vote(
+        &self,
+        request: Request<RawSignVoteRequest>,
+    ) -> Result<Response<RawSignedVoteResponse>, Status> {
+        let req = SignVoteRequest::try_from(request.into_inner())
+            .map_err(|e| Status::invalid_argument(e.to_string()))?;
+        debug!(
+            "[{}] received a vote signing request: {:?}",
+            req.chain_id, req
+        );
+        self.check_max_height(&req.chain_id, req.vote.height)
+            .map_err(|_| Status::failed_precondition("max height exceeded"))?;
+        let mut providers = self.providers.lock().await;
+        let resp = if let Some((signer, state_provider)) = providers.get_mut(&req.chain_id) {
```

Comment:
Seems like this would be a bit more readable if you hoisted the error handling which is currently on [L269](https://github.com/informalsystems/tendermint-rs/pull/1152/files#diff-beb011520d9feeec8b734a23ccbae4e32ba3c5e635a23970bb620fbe20a3cd68R269) up here.